### PR TITLE
Add emitter framework

### PR DESCRIPTION
**Pull request by bterlson**
This PR implements a framework that makes writing emitters substantially easier. For details, see resources below.

Pending API review before checking in.

### Useful resources

* [Conceptual docs](https://cadlwebsite.z1.web.core.windows.net/prs/1528/next/extending-cadl/emitter-framework.html)

### Todo

- [x] Change asserts to diagnostics.
- [x] Generate library documentation, add documentation where needed (approx. everywhere).
- [x] End-to-end testing (remove console log based testing)

### CODE REVIEW

**Review by allenjzhang — APPROVED** [review]

**Comment by nguerrera** [comment]
> the link doesn't seem to work

FYI. This is a weird quirk of our staging hosting where you have to append .html to Try-It URLs:

https://cadlwebsite.z1.web.core.windows.net/prs/1528/next/extending-cadl/emitter-framework.html

Anyway, reading through this now :)

**Comment by bterlson** [comment]
Alright, most of the implementation code is landed. I added a checklist of things that need to happen still, but feel free to review the code at this point!

File: packages/compiler/emitter-framework/types.ts
Comment (allenjzhang) [inline packages/compiler/emitter-framework/types.ts:15]: nit: /** */.

**Comment by bterlson** [comment]
It turns out we don't yet have API Docs for compiler APIs. As such, I'm backburnering API docs for now, and focusing on other things.

File: packages/compiler/emitter-framework/types.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/types.ts:15]: Leaving many of these docs until later since I want to iterate on them while looking at the actual output

**Comment by bterlson** [comment]
CI is green, but I will wait for code review before merging. I'll have limited bandwidth this week to respond to feedback!

**Comment by bterlson** [comment]
Ahh I realize that a great section to add to the conceptual docs is how to test the emitter. I will work on adding that later.

File: docs/extending-cadl/emitter-framework.md
Comment (nguerrera) [inline docs/extending-cadl/emitter-framework.md:46]:
```suggestion
This is the base class for writing logic to convert Cadl types into assets in your target language. Every Cadl type has at least one method on this base class, and many have multiple methods. For example, models have both `ModelDeclaration` and `ModelLiteral` methods to handle `model Pet { }` declarations and `{ anonymous: boolean }` literals respectively.
```

File: packages/compiler/emitter-framework/custom-key-map.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/custom-key-map.ts:20]: Can this/should this replace MultiKeyMap or should MultiKeyMap be built on top of this?

File: packages/compiler/emitter-framework/builders/array.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/builders/array.ts:5]: Nit (feel free to won't fix): Consider putting the builders together in a single file like builders.ts or renaming the files to include builder. I think array.ts/object.ts/string.ts is likely to trip me up when using Go-to-file.

File: packages/compiler/emitter-framework/asset-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/asset-emitter.ts:115]: Shouldn't these doc comments be moved to the interface so that consumers see it in the IDE?

File: packages/compiler/emitter-framework/asset-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/asset-emitter.ts:666]: worth assert never trick here to catch at compile time?

File: packages/compiler/emitter-framework/asset-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/asset-emitter.ts:696]:
```suggestion
 * This implementation is EXTREMELY non-optimal (O(n*m) where n = number of unique
```

File: packages/compiler/emitter-framework/asset-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/asset-emitter.ts:703]: Worth filing a tracking bug to optimize or do you think it is OK to wait for the bottleneck to arise in practice?

File: packages/compiler/emitter-framework/type-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/type-emitter.ts:40]: Should it be marked `abstract` then?

File: packages/compiler/emitter-framework/type-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/type-emitter.ts:213]: nit: Consider removing empty @returns tags.

File: packages/compiler/emitter-framework/type-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/type-emitter.ts:258]: should this one have a return type other than inferred `{}`?

File: packages/compiler/emitter-framework/type-emitter.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/type-emitter.ts:642]:
```suggestion
    // operations inside instantiated interfaces get weird names
```

File: packages/compiler/emitter-framework/types.ts
Comment (nguerrera) [inline packages/compiler/emitter-framework/types.ts:75]: Is it going to cause problems that there is also a SourceFile core type? Should we pick another name?

**Review by nguerrera — APPROVED** [review]
This is amazing work! Thank you for writing detailed documentation and implementation comments. It sets a great example to follow. I've made two passes over the docs and change now and only found minor things above. 

I'm eager to try this out. I struggled with some sample code that got obfuscated by having to deal with things that the framework provides. Given the size of the change, it's likely that I've missed some things, and I will probably end up with more questions and better insights as I use it.

File: packages/compiler/emitter-framework/asset-emitter.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/asset-emitter.ts:703]: I would guess this is in the bucket of things that are slow but unlikely to show up as important behind other slow things except for severely degenerate cases that may not show up in practice.

Comment (bterlson) [reply packages/compiler/emitter-framework/asset-emitter.ts:703]: IOW, I would say let's not track. Just something to keep in mind!

File: packages/compiler/emitter-framework/types.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/types.ts:75]: Yeah, I'll rename to like EmittedSourceFile or so.

Comment (bterlson) [reply packages/compiler/emitter-framework/types.ts:75]: (Technically it won't cause a conflict since the emitter framework is exported under a separate entrypoint, but better not to cause confusion).

File: packages/compiler/emitter-framework/custom-key-map.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/custom-key-map.ts:20]: It could, I suppose, but the interface is quite a bit different. Objections to leaving this for later?

File: packages/compiler/emitter-framework/asset-emitter.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/asset-emitter.ts:115]: I've deleted these docs... I will make a documentation PR in a subsequent pass (want to iterate on the docs once the compiler docs get checked in).

File: packages/compiler/emitter-framework/asset-emitter.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/asset-emitter.ts:654]: `type.kind` isn't never here, there are some things that shouldn't show up in the type graph that are nonetheless types (`Function`, `TypeParameter`, etc.)

File: packages/compiler/emitter-framework/type-emitter.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/type-emitter.ts:40]: It turns out the answer to this is no, because it causes significant type complexity having to do with the fact that functions which take a `typeof TypeEmitter` as a parameter (like createAssetEmitter)... can't anymore, because the compiler complains that it's abstract. The general fix (I THINK) is to replace such instances with `new<T>(assetEmitter: AssetEmitter<T>): TypeEmitter<T>`, but that was having deleterious effects all over the codebase. I'm giving up.

File: packages/compiler/emitter-framework/custom-key-map.ts
Comment (nguerrera) [reply packages/compiler/emitter-framework/custom-key-map.ts:20]: No objections, it seemed there was *some* overlap and I was mostly curious.

**Pull request merged**
